### ISSUE TITLE
Distorted datamatrix codes fail to decode

### ISSUE THREAD

**Issue by axxel** [nu-book/zxing-cpp#37]
@bommo1 came across a set of datamatrix samples that fail to decode because they are distorted. The larger the codes, the more likely will happen. Think: lens distortion, or not completely flat paper.

This issue is supposed to collect information about possible improvements.

One way to potentially solve the problem is by parsing the alignment patterns inside large codes (that is what they are there for anyway) and then sample each sub code with a regular grid. If the distortion relative to the module size is too big, that will also fail, though. For the provided samples below, this might have a chance.

Here are the samples:

![1](https://user-images.githubusercontent.com/32093209/36216184-55fc71c4-11ae-11e8-800b-2d1fc71b9a42.png)

![4](https://user-images.githubusercontent.com/32093209/36250731-f2187c20-123e-11e8-9015-25f51a815417.png)

![medikationsplan1](https://user-images.githubusercontent.com/1038346/36309272-47f29c80-1324-11e8-9b02-a46fee5d7081.png)

**Comment by huycn** [comment]
We may also deskew the image beforehand. It can easily be done with the help of library such as OpenCV. What do you think?

**Comment by bommo1** [comment]
@huycn it's easy once you know where the distorted grid or at least its intersection points are. Hough transformation and the Line segment detector do not find noisy lines. Is there a simpler way that I'm missing? Using only the contour might not be enough.

**Comment by huycn** [comment]
You are right, I overlooked. Here what we got is curves not lines, thus it's not that easy. But when you tried Hough transformation, why did it fail? Will it help if divide the picture into smaller areas and do Hough transforms on each of them?

**Comment by axxel** [comment]
In my opinion, re-sampling the source image to fix a distortion makes sense if the distortion is a lens artifact. Calibrating lens distortion is pretty much off-the-shelf stuff (e.g. in OpenCV). If the distortions come from a bent/wavy peace of paper the easiest solution would be to flatten it, by e.g. pressing it on a peace of glass. If that is not an option, then first analyzing the code to come up with some kind of control point set for some inverse distortion function from OpenCV to then binarize/scan it a second time seems overly complicated plus a performance pessimization compared to what I outlined above.

**Comment by huycn** [comment]
I agree with you on performance impact. It seems to be more a domain issue and must be treated as such. If the library is used in un-interactive process, the pictures can be processed according to domain's specifics. If it is used in an interactive mobile app like ours, it's not an issue in the first place.

**Comment by bommo1** [comment]
Problematic are, especially in the interactive process, big data matrix codes like the 3rd example from above that are printed on a small area. When the camera is far then there's no noticeable distortion but because the camera is in video mode the resolution is too small to be able to read the code. When you get closer lens distortion kicks in.
I'll try to un-distort using only the contour. It should help in some cases.

**Comment by bommo1** [comment]
I had some mild success with the reshaping. It would be nice though if the detector would work with minor distortions.

**Comment by bommo1** [comment]
The problem with the big DM Code (3rd example) is not only the distortion. Error check/correction (RS) fails. If I comment these 2 lines in the DMDecoder I seem to get the right value back.

if (!CorrectErrors(codewordBytes, numDataCodewords))
			return DecodeStatus::ChecksumError;

I can't tell if the DM-code is wrong or if the decoder is. From 6 software/libs I've tried 3 can read it and 3 can not.

![frame_00000_med](https://user-images.githubusercontent.com/32093209/37153710-efe509fa-22dd-11e8-94bb-bf42e5bfbaf0.png)

**Comment by bommo1** [comment]
I had some time to come back to the problematic DM-code from above.
It turns out that issue was caused not by the ReedSolomonDecoder but by the way the error code-words get de-interleaved.

If you change this line https://github.com/nu-book/zxing-cpp/blob/master/core/src/datamatrix/DMDataBlock.cpp#L68 to `int jOffset = j;` the RSDecoder is happy for this particular code but not for other 144x144 codes found on the internet.

I've found two software that can read both types of codes:
https://online-barcode-reader.inliteresearch.com
https://bytescout.com/products/enduser/misc/barcodereader.html

As I don't have access to the data matrix specification (ISO 16022) I see two possible situations:
1) the DM specification allows two types of interleaving but zxing supports only one. Not very probable.
2) there are DM generators in the wild that do not implement the specification correctly and use interleaving without offset. Some libraries handle this bug and are able to read invalid codes.

Either way this change might make sense for others too:
DMDataBlock.GetDataBlocks() - support both de-interleaving methods
DMDecoder.Decode() - fallback to the second method when the RS-Decoder is unhappy

**Comment by bommo1** [comment]
Looking at the table below it looks more like zxing is doing it wrong.

![iso16022](https://user-images.githubusercontent.com/32093209/46721554-9224f100-cc73-11e8-9f41-5abf3bdb536b.png)

The behavior was changed here:
https://github.com/zxing/zxing/issues/262
https://github.com/zxing/zxing/commit/344d55e6889875ab23d201fb16b5ef5993912f55

**Comment by axxel** [comment]
Thanks for digging into this. I tried to follow your investigation but I'm not sure I got the full picture. Does this align with your understanding:
 1. zxing used to do the 'right' thing but someone found it fails to decode a bunch of 'real world' codes (issue 262).
 1. zxing was changed, so it decoded those samples, thereby breaking 'correct' codes (like the one you have now).
 1. the two different versions of ECC interlacing can not be distinguished.
 1. the postscript barcode guy found presumably the 'same' bug and it fixed it? for both types? how?

Does that mean there is a ISO specification and a bunch of generators clearly violate that spec? Do you have a suggestion for how to fix this?

**Comment by bommo1** [comment]
There's the "correct way" according to the specification and then there is the "convention in the industry". These do not match and there's no way to distinguish between them by looking at the DM code.

zxing currently implements the "convention in the industry". It used to implement the "correct way" (before this commit zxing/zxing@344d55e).

Some other barcode scanning software take the safe path and support both (as fallback). This is reasonable thing to do.

The problem is what to do with the generator. The "convention in the industry"-way seems to be supported by most scanners out there whereas the "correct way" is not.

**Comment by bommo1** [comment]
https://github.com/bwipp/postscriptbarcode/blob/master/src/datamatrix.ps#L773

Hard to tell but It looks like "convention in the industry"-only.
